### PR TITLE
Add `test` Jinja tag

### PR DESCRIPTION
**Pull request by kwigley**
addresses part of #1173

### Description

This PR adds a `test` Jinja tag. This gives us the ability to treat tests differently than very specifically defined macros. There is no change in behavior with this PR.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: core/dbt/compilation.py [hunk 1]
```diff
@@ -250,9 +250,6 @@ def _inject_ctes_into_sql(self, sql: str, ctes: List[InjectedCTE]) -> str:
 
         return str(parsed)
 
-    def _get_dbt_test_name(self) -> str:
```
Comment (kwigley) [inline core/dbt/compilation.py:253]: Wee, since we don't prepend tests with `count(*)` when "compiling" this can go away!

**Review by iknox-fa — APPROVED** [review]

File: core/dbt/utils.py [hunk 2]
```diff
@@ -101,6 +101,15 @@ def get_docs_macro_name(docs_name, with_prefix=True):
         return docs_name
 
 
+def get_test_macro_name(test_name, with_prefix=True):
+    name = f'test_{test_name}'
+
+    if with_prefix:
```
Comment (iknox-fa) [inline core/dbt/utils.py:107]: Nit:  Use ternary here?

**Review by iknox-fa — APPROVED** [review]

**Review by jtcohen6 — APPROVED** [review]
This is backwards compatible in a very clever way, and I had to spend a few minutes thinking about it to really get it. Nice work!

A few things worth calling out, either because I think we should eventually document them, or because it reorients my own thinking:
- Test blocks are still technically macro nodes, just like materializations. As long as that's true, some of our other work around "configuring" generic tests (e.g. from `dbt_project.yml`) will have to work a bit differently. I actually think that's ok!
- A test node named `my_test` cannot coexist with a macro named `test_my_test`. I suppose the same is true of a materialization named `my_mtzn` and a macro named `materialization_my_mtzn`.
- If users are commonly calling other macros from within a test block, it's more likely that they will encounter the v0.19.1 regression (`config()` block or `ref` not properly accounted for), which @gshank is fixing by sticking the top-level test macro into the capture context. Once more, I think that's ok: We can tell users they need to hoist ref/source/config within the main test block to achieve the functionality they're after.

**Review by gshank — APPROVED** [review]

**Review by nathaniel-may — APPROVED** [review]
This is very slick! thanks for adding all the comments 🥇

**Pull request merged**